python/python/lance/ray/sink.py

```diff
@@ -0,0 +1,133 @@
+# SPDX-License-Identifier: Apache-2.0
+# SPDX-FileCopyrightText: Copyright The Lance Authors
+
+import logging
+from typing import TYPE_CHECKING, Iterable, Optional, Union
+
+import pyarrow as pa
+
+import lance
+from lance.fragment import FragmentMetadata, LanceFragment
+
+from ..dependencies import ray
+
+if TYPE_CHECKING:
+    import pandas as pd
+
+__all__ = ["LanceDatasink"]
+
+
+class LanceDatasink(ray.data.Datasink):
+    """Lance Ray Datasink.
+
+    Write a Ray dataset to lance.
+
+    Parameters
+    ----------
+    uri : str
+        The base URI of the dataset.
+    schema : pyarrow.Schema
+        The schema of the dataset.
+    mode : str, optional
+        The write mode. Default is 'append'. Choices are 'append' and 'overwrite'.
+    max_rows_per_file : int, optional
+        The maximum number of rows per file. Default is 1024 * 1024.
+
+
+    Examples
+    --------
+
+    .. code-block:: python
+
+        import ray
+        import lance
+        import pyarrow as pa
+
+        ray.init()
+
+        schema = pa.schema([pa.field("id", pa.int64()), pa.field("str", pa.string())])
+        sink = lance.ray.sink.LanceDatasink("path/to/dataset", schema=schema)
+        (
+            ray.data
+            .range(10)
+            .map(lambda x: {"id": x["id"], "str": f"str-{x['id']}"})
+            .write_datasink(sink)
+        )
+
+    """
+
+    NAME = "Lance"
+
+    def __init__(
+        self,
+        uri: str,
+        schema: Optional[pa.Schema] = None,
+        mode: str = "create",
+        max_rows_per_file: int = 1024 * 1024,
+        *args,
+        **kwargs,
+    ):
+        super().__init__(*args, **kwargs)
+
+        self.uri = uri
+        self.schema = schema
+        self.mode = mode
+
+        if self.mode in set(["create", "overwrite"]) and self.schema is None:
+            raise ValueError("schema is required for create and overwrite modes")
```

Comment:
Not a big deal but is there any way we can determine the schema from the first batch of data?

Reply:
Add a `Literal[]` for type annotation.